samplejava/src/main/java/com/example/chattutorialjava/ChannelActivity3.java

```diff
@@ -45,100 +45,91 @@ public static Intent newIntent(Context context, Channel channel) {
+        messageListViewModel.getState().observe(this, state -> {
+            if (state instanceof NavigateUp) {
+                finish();
             }
-            channelHeaderView.setText(typing);
->>>>>>> 23067e524310df9e383d921476e1f7625c2b04df
         });
 
-        final MessageInputViewModel messageInputViewModel = viewModelProvider.get(MessageInputViewModel.class);
-        MessageInputViewModelBinding.bind(messageInputViewModel, messageInputView, this);
-        messageListViewModel.getMode().observe(this, mode -> {
-            if (mode instanceof MessageListViewModel.Mode.Thread) {
-                messageInputViewModel.setActiveThread(((MessageListViewModel.Mode.Thread) mode).getParentMessage());
-            } else if (mode instanceof MessageListViewModel.Mode.Normal) {
-                messageInputViewModel.resetThread();
-            }
-            messageListView.setOnMessageEditHandler((message) -> {
-                messageInputViewModel.getEditMessage().postValue(message);
-                return Unit.INSTANCE;
-            });
+        // Step 5 - Let the message input know when we are editing a message
+        messageListView.setOnMessageEditHandler(message -> {
+            messageInputViewModel.getEditMessage().postValue(message);
+            return Unit.INSTANCE;
         });
 
-        Function0<Unit> onBackClick = () -> {
+        // Step 6 - Handle back button behaviour correctly when you're in a thread
+        channelHeaderView.setOnBackClick(() -> {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed.INSTANCE);
             return Unit.INSTANCE;
-        };
-
+        });
         getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
             @Override
             public void handleOnBackPressed() {
-                onBackClick.invoke();
+                channelHeaderView.getOnBackClick().invoke();
+            }
+        });
+
+        // Custom typing info header bar
+        String nobodyTyping = "nobody is typing";
+        typingHeader.setText(nobodyTyping);
+
+        // Obtain a ChannelController
+        ChannelController channelController =
+                ChatDomain.instance().getUseCases()
+                        .getWatchChannel().invoke(cid, 0)
+                        .execute().data();
+
+        // Observe typing users
+        channelController.getTyping().observe(this, users -> {
+            if (users.isEmpty()) {
+                typingHeader.setText(nobodyTyping);
+            } else {
+                StringBuilder typingText = new StringBuilder("typing: ");
+                for (int i = 0; i < users.size(); i++) {
+                    User user = users.get(i);
+                    if (i > 0) {
+                        typingText.append(", ");
+                    }
+                    typingText.append(user.getExtraData().get("name"));
+                }
+                typingHeader.setText(typingText.toString());
```

Comment:
`joinToString` is a godsend, this was way too difficult to implement in Java

Reply:
Agree, had the same problem some time ago 🙈

Reply:
Could maybe cheat and use CollectionsKt.joinToString(...) :D

Reply:
You actually can't - it would look really, really ugly, as you'd have to provide _all_ of its parameters. It has a lot of optional parameters.